// prisma-db-introspection/dist/common/modelNameAndDirectiveNormalizer.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var modelNameNormalizer_1 = require("./modelNameNormalizer");
var ModelNameAndDirectiveNormalizer = /** @class */ (function (_super) {
    __extends(ModelNameAndDirectiveNormalizer, _super);
    function ModelNameAndDirectiveNormalizer(baseModel) {
        var _this = _super.call(this) || this;
        _this.baseModel = baseModel;
        _this.baseType = null;
        return _this;
    }
    // https://github.com/prisma/prisma/issues/3725
    ModelNameAndDirectiveNormalizer.prototype.normalize = function (model) {
        _super.prototype.normalize.call(this, model);
    };
    ModelNameAndDirectiveNormalizer.prototype.findBaseByName = function (baseObjs, obj) {
        var _a = __read(baseObjs.filter(function (base) {
            if (base.databaseName) {
                return base.databaseName === obj.name;
            }
            else {
                return base.name === obj.name;
            }
        }), 1), baseCandidate = _a[0];
        return baseCandidate || null;
    };
    ModelNameAndDirectiveNormalizer.prototype.findBaseById = function (baseObjs, obj) {
        var _a = __read(baseObjs.filter(function (baseObj) { return baseObj.isId && obj.isId; }), 1), baseCandidate = _a[0];
        return baseCandidate || null;
    };
    ModelNameAndDirectiveNormalizer.prototype.assignProperties = function (baseObj, obj) {
        if (baseObj.databaseName) {
            obj.name = baseObj.name;
            obj.databaseName = baseObj.databaseName;
        }
    };
    ModelNameAndDirectiveNormalizer.prototype.assignTypeProperties = function (baseObj, obj) {
        if (baseObj === null) {
            return;
        }
        this.assignProperties(baseObj, obj);
    };
    ModelNameAndDirectiveNormalizer.prototype.assignFieldProperties = function (baseObj, obj) {
        if (baseObj === null) {
            return;
        }
        this.assignProperties(baseObj, obj);
        obj.isId = obj.isId || baseObj.isId;
        obj.isCreatedAt = obj.isCreatedAt || baseObj.isCreatedAt;
        obj.isUpdatedAt = obj.isUpdatedAt || baseObj.isUpdatedAt;
    };
    ModelNameAndDirectiveNormalizer.prototype.normalizeType = function (type) {
        this.baseType = this.baseModel === null ? null : this.findBaseByName(this.baseModel.types, type);
        this.assignTypeProperties(this.baseType, type);
        _super.prototype.normalizeType.call(this, type);
    };
    ModelNameAndDirectiveNormalizer.prototype.normalizeField = function (field, parentType) {
        var baseField = null;
        if (this.baseType !== null) {
            baseField = this.findBaseByName(this.baseType.fields, field);
            if (baseField === null) {
                // Fallback - if no name match, we check if the field is an ID.
                baseField = this.findBaseById(this.baseType.fields, field);
                // Special handling for ID fields. We just assign the name.
                if (baseField !== null) {
                    this.assignName(field, baseField.name);
                }
            }
        }
        this.assignFieldProperties(baseField, field);
        _super.prototype.normalizeField.call(this, field, parentType);
    };
    return ModelNameAndDirectiveNormalizer;
}(modelNameNormalizer_1.default));
exports.default = ModelNameAndDirectiveNormalizer;
